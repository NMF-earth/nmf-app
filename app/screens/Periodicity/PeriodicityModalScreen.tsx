import React, { useState } from "react";
import { ScrollView, View } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { pathOr } from "ramda";
import { useRoute } from "@react-navigation/core";

import { NavStatelessComponent, PeriodicityType } from "interfaces";
import { Button, Text } from "components";
import ClickableTag from "components/ClickableTag";
import { TranslationKeys } from "utils/translations/resources";
import { t } from "utils";

import styles from "./PeriodicityModalScreen.styles";

const DEFAULT_VALUES = {
  periodType: PeriodicityType.monthly,
  weekDays: [],
  periodTimes: 0,
};

const getPeriodType = pathOr(DEFAULT_VALUES.periodType, ["params", "periodType"]);
const getWeekDays = pathOr(DEFAULT_VALUES.weekDays, ["params", "periodWeekDays"]);
const getTimes = pathOr(DEFAULT_VALUES.periodTimes, ["params", "periodTimes"]);

export const PeriodicityModalScreen: NavStatelessComponent = () => {
  const route = useRoute();
  const navigation = useNavigation();

  const [periodType, setPeriodType] = useState<PeriodicityType>(getPeriodType(route));
  const [weekDays, setWeekdays] = useState<number[]>(getWeekDays(route));
  const [periodTimes, setTimes] = useState<number>(getTimes(route));

  function onWeekdaySelected(dayIndex: number) {
    setWeekdays(
      weekDays.includes(dayIndex) ? weekDays.filter((i) => i !== dayIndex) : [...weekDays, dayIndex]
    );
  }

  function onConfirm() {
    navigation.navigate("AddEmission", {
      periodType,
      periodWeekDays: periodType === PeriodicityType.weekly ? weekDays : null,
      periodTimes,
    });
  }

  function onCancel() {
    navigation.navigate("AddEmission", {
      periodType: DEFAULT_VALUES.periodType,
      periodWeekDays: DEFAULT_VALUES.weekDays,
      periodTimes: DEFAULT_VALUES.periodTimes,
    });
  }

  const hideConfirm =
    !periodType || !periodTimes || (periodType === PeriodicityType.weekly && weekDays.length === 0);

  return (
    <ScrollView style={styles.container}>
      <View style={styles.tagSection}>
        <Text.H3>{t("PERIODICITY_MODAL_SCREEN_OFTEN")}</Text.H3>
        <View style={styles.tagsContainer}>
          <ClickableTag
            isSelected={periodType === PeriodicityType.monthly}
            onPress={() => {
              setPeriodType(PeriodicityType.monthly);
            }}
            text={t("PERIODICITY_MODAL_SCREEN_MONTHLY")}
          />

          <ClickableTag
            isSelected={periodType === PeriodicityType.weekly}
            onPress={() => {
              setPeriodType(PeriodicityType.weekly);
            }}
            text={t("PERIODICITY_MODAL_SCREEN_WEEKLY")}
          />
        </View>
      </View>
      {periodType === PeriodicityType.weekly && (
        <View style={styles.tagSection}>
          <Text.H3>{t("PERIODICITY_MODAL_SCREEN_DAYS")}</Text.H3>
          <View style={styles.tagsContainer}>
            {WEEK_DAYS_LIST.map(({ dayIndex, nameKey }) => {
              return (
                <ClickableTag
                  key={dayIndex}
                  isSelected={weekDays.includes(dayIndex)}
                  onPress={() => {
                    onWeekdaySelected(dayIndex);
                  }}
                  text={t(nameKey)}
                />
              );
            })}
          </View>
        </View>
      )}
      <View style={styles.tagSection}>
        <Text.H3>{t("PERIODICITY_MODAL_SCREEN_OCCURRENCES")}</Text.H3>
        <View style={styles.tagsContainer}>
          {TIMES_LIST.map((_, index) => (
            <ClickableTag
              key={index}
              isSelected={periodTimes === index + 1}
              onPress={() => {
                setTimes(index + 1);
              }}
              text={`${index + 1} ${
                index > 0 ? t("PERIODICITY_MODAL_SCREEN_TIMES") : t("PERIODICITY_MODAL_SCREEN_TIME")
              }`}
            />
          ))}
        </View>
      </View>

      {!hideConfirm && (
        <View style={styles.confirmBtnContainer}>
          <Button.Primary
            fullWidth
            onPress={onConfirm}
            text={t("PERIODICITY_MODAL_SCREEN_COMFIRM")}
          />
          <View style={{ height: 12 }} />
          <Button.Secondary
            fullWidth
            onPress={onCancel}
            text={t("PERIODICITY_MODAL_SCREEN_CANCEL_NO_PERIODICITY")}
          />
        </View>
      )}
    </ScrollView>
  );
};

const TIMES_LIST = new Array(3).fill(null);

export const WEEK_DAYS_LIST: {
  dayIndex: number;
  nameKey: keyof TranslationKeys;
}[] = [
  { dayIndex: 1, nameKey: "UI_SUNDAY" },
  { dayIndex: 2, nameKey: "UI_MONDAY" },
  { dayIndex: 3, nameKey: "UI_TUESDAY" },
  { dayIndex: 4, nameKey: "UI_WEDNESDAY" },
  { dayIndex: 5, nameKey: "UI_THURSDAY" },
  { dayIndex: 6, nameKey: "UI_FRIDAY" },
  { dayIndex: 7, nameKey: "UI_SATURDAY" },
];
