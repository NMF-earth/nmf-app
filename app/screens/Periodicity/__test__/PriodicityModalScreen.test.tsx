import React from "react";
import { create } from "react-test-renderer";

import { PeriodicityModalScreen } from "../PeriodicityModalScreen";

it("InfoModalScreen renders correctly", () => {
  const tree = create(<PeriodicityModalScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
