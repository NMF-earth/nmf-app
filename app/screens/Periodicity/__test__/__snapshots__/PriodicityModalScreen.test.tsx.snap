// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoModalScreen renders correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "paddingHorizontal": 16,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "marginVertical": 5,
        }
      }
    >
      <Text.H3>
        How often this emission happens?
      </Text.H3>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "marginVertical": 5,
          }
        }
      >
        <TouchableOpacity
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "center",
                "borderRadius": 6,
                "borderWidth": 1.5,
                "margin": 6,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
              Object {
                "backgroundColor": "#DCF0E5",
                "borderColor": "#4BB377",
              },
              undefined,
            ]
          }
        >
          <Text.Secondary
            bold={true}
            darkGray={false}
            green={true}
          >
            Monthly
          </Text.Secondary>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "center",
                "borderRadius": 6,
                "borderWidth": 1.5,
                "margin": 6,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
              Object {
                "backgroundColor": "#F0F0F0",
                "borderColor": "#9CA6A0",
              },
              undefined,
            ]
          }
        >
          <Text.Secondary
            bold={true}
            darkGray={true}
            green={false}
          >
            Weekly
          </Text.Secondary>
        </TouchableOpacity>
      </View>
    </View>
    <View
      style={
        Object {
          "marginVertical": 5,
        }
      }
    >
      <Text.H3>
        Occurences?
      </Text.H3>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "marginVertical": 5,
          }
        }
      >
        <TouchableOpacity
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "center",
                "borderRadius": 6,
                "borderWidth": 1.5,
                "margin": 6,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
              Object {
                "backgroundColor": "#F0F0F0",
                "borderColor": "#9CA6A0",
              },
              undefined,
            ]
          }
        >
          <Text.Secondary
            bold={true}
            darkGray={true}
            green={false}
          >
            1 time
          </Text.Secondary>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "center",
                "borderRadius": 6,
                "borderWidth": 1.5,
                "margin": 6,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
              Object {
                "backgroundColor": "#F0F0F0",
                "borderColor": "#9CA6A0",
              },
              undefined,
            ]
          }
        >
          <Text.Secondary
            bold={true}
            darkGray={true}
            green={false}
          >
            2 times
          </Text.Secondary>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "center",
                "borderRadius": 6,
                "borderWidth": 1.5,
                "margin": 6,
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
              Object {
                "backgroundColor": "#F0F0F0",
                "borderColor": "#9CA6A0",
              },
              undefined,
            ]
          }
        >
          <Text.Secondary
            bold={true}
            darkGray={true}
            green={false}
          >
            3 times
          </Text.Secondary>
        </TouchableOpacity>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
